---
title: "Utilizing LIGER for the integration of spatial transcriptomic data"
author: "Joshua Sodicoff"
date: "3/6/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(liger)
library(magrittr)
library(ggplot2)
ggplot2::theme_set(theme_cowplot())
load("birs_data.RData")
metadata = read.csv("GSE115746_complete_metadata_28706-cells.csv")
samples = intersect(colnames(tasic), as.character(metadata[as.character(metadata$source_name) == "Primary Visual Cortex (VISp)",]$sample_name))
tasic = tasic[,samples]
tasic_clust =tasic_clust[samples]
```

To increase the resolution of the scRNAseq, we utilized the dataset from [Tasic et al](https://www.nature.com/articles/s41586-018-0654-5), with over 43,000 genes and 23,000 cells.

```{r seqfish_int, warning = F, message = F, results = "hide"}
liger_seqfish = createLiger(list(tasic = tasic, seqfish = seqfish))
liger_seqfish %<>% normalize() %>% selectGenes() %>% scaleNotCenter()
liger_seqfish %<>% optimizeALS(k=20) %>% quantile_norm() %>% louvainCluster()
liger_seqfish %<>% runTSNE()
names(liger_seqfish@clusters)=rownames(liger_seqfish@tsne.coords)
seqfish_purity = calcPurity(liger_seqfish, seq_clust)
tasic_purity = calcPurity(liger_seqfish, tasic_clust)
seqfish_ari = calcARI(liger_seqfish, seq_clust)
tasic_ari = calcARI(liger_seqfish, tasic_clust)
alignment = calcAlignment(liger_seqfish)
seqfish_stats = c(seqfish_purity, seqfish_ari, tasic_purity, tasic_ari, alignment)
by_dataset = plotByDatasetAndCluster(liger_seqfish, return.plots = T)[[1]]
by_liger_cluster = plotByDatasetAndCluster(liger_seqfish, return.plots = T)[[2]]
by_tasic_cluster = plotByDatasetAndCluster(liger_seqfish, clusters = tasic_clust, return.plots = T)[[2]]
by_seqfish_cluster = plotByDatasetAndCluster(liger_seqfish, clusters = seq_clust, return.plots = T)[[2]]
```
```{r display_seqfish, warning = F}
seqfish_stats
by_dataset
by_liger_cluster
by_tasic_cluster
by_seqfish_cluster
```

To directly compare the performance on the given seqfish data to a baseline dataset, the 160 gene STARmap dataset utilized in [Wang and Allen et. al](https://www.dropbox.com/sh/f7ebheru1lbz91s/AAC-QLN6MqkbrwY6XzpRj-soa/visual_160?dl=0&subfolder_nav_tracking=1) was downsampled to 120 genes and integrated with the scRNAseq dataset. 

```{r 120_downsampled, warning = F, message = F, results = "hide"}
starmap_downsampled = starmap_160[sample(1:160, 120, replace = FALSE),]
liger_120 = createLiger(list(tasic = tasic, starmap = starmap_downsampled))
liger_120 %<>% normalize() %>% selectGenes() %>% scaleNotCenter()
liger_120 %<>% optimizeALS(k=20) %>% quantile_norm() %>% louvainCluster()
liger_120 %<>% runTSNE()
names(liger_120@clusters)=rownames(liger_120@tsne.coords)
starmap_purity = calcPurity(liger_120, starmap_160_clust)
tasic_purity = calcPurity(liger_120, tasic_clust)
starmap_ari = calcARI(liger_120, starmap_160_clust)
tasic_ari = calcARI(liger_120, tasic_clust)
alignment = calcAlignment(liger_120)
starmap_stats = c(seqfish_purity, seqfish_ari, tasic_purity, tasic_ari, alignment)
by_dataset = plotByDatasetAndCluster(liger_120, return.plots = T)[[1]]
by_liger_cluster = plotByDatasetAndCluster(liger_120, return.plots = T)[[2]]
by_tasic_cluster = plotByDatasetAndCluster(liger_120, clusters = tasic_clust, return.plots = T)[[2]]
by_starmap_cluster = plotByDatasetAndCluster(liger_120, clusters = starmap_clust, return.plots = T)[[2]]
```

```{r display_120, warning = F}
starmap_stats
by_dataset
by_liger_cluster
by_tasic_cluster
by_starmap_cluster
```
```{r stats_comp, echo = F}
dataset = c(rep("seqfish", 5),rep("starmap", 5))
stat_type = rep(c( "spatial_purity","spatial_ari","tasic_purity","tasic_ari","alignment"), 2)
value = c(seqfish_stats, starmap_stats)
summary = data.frame(dataset, stat_type, value)
ggplot(summary, aes(fill = dataset, y = value, x=stat_type))+geom_bar(position="dodge", stat="identity")
```

To determine the impact of the number of genes in spatial data, a dataset with a larger number of genes was utilized. The 1020 gene STARmap dataset utilized in [Wang and Allen et. al](https://www.dropbox.com/sh/f7ebheru1lbz91s/AAC3l8TZSNhSWmfILw-2dglma/visual_1020/20180410-BY3_1kgenes?dl=0&subfolder_nav_tracking=1) was repeatedly downsampled and integrated with the much larger scRNAseq dataset.

```{r 700_downsampled, warning = F, message = F, results = "hide"}
starmap_downsampled = starmap_1020[sample(1:1020, 700, replace = FALSE),]
liger_700 = createLiger(list(tasic = tasic, starmap = starmap_downsampled))
liger_700 %<>% normalize() %>% selectGenes() %>% scaleNotCenter()
liger_700 %<>% optimizeALS(k=20) %>% quantile_norm() %>% louvainCluster()
liger_700 %<>% runTSNE()
names(liger_700@clusters)=rownames(liger_700@tsne.coords)
starmap_purity = calcPurity(liger_700, starmap_clust)
tasic_purity = calcPurity(liger_700, tasic_clust)
starmap_ari = calcARI(liger_700, starmap_clust)
tasic_ari = calcARI(liger_700, tasic_clust)
alignment = calcAlignment(liger_700)
by_dataset = plotByDatasetAndCluster(liger_700, return.plots = T)[[1]]
by_liger_cluster = plotByDatasetAndCluster(liger_700, return.plots = T)[[2]]
by_tasic_cluster = plotByDatasetAndCluster(liger_700, clusters = tasic_clust, return.plots = T)[[2]]
by_starmap_cluster = plotByDatasetAndCluster(liger_700, clusters = starmap_clust, return.plots = T)[[2]]
```

```{r display_700, warning = F}
c(starmap_purity, starmap_ari, tasic_purity, tasic_ari, alignment)
by_dataset
by_liger_cluster
by_tasic_cluster
by_starmap_cluster
```

To account for the randomness associated with downsampling genes and the stochastic nature of our method, we repeated these analyses several times and took an average across runs.

```{r rep, eval = F}
stats_starmap = matrix(ncol = 6)
colnames(stats_starmap) = c("genes","purity_spatial","purity_rnaseq","ari_spatial","ari_rnaseq","alignment")
stats_seqfish = stats_starmap
for(n in 1:10){
  for(i in seq(900, 100, -100)){
      starmap_downsampled = starmap_1020[sample(1:1020, i, replace = FALSE),]
      liger_down = createLiger(list(tasic = tasic, starmap = starmap_downsampled))
      liger_down %<>% normalize() %>% selectGenes() %>% scaleNotCenter()
      liger_down %<>% optimizeALS(k=20) %>% quantile_norm() %>% louvainCluster()
      liger_down %<>% runTSNE()
      names(liger_down@clusters)=rownames(liger_down@tsne.coords)
      starmap_purity = calcPurity(liger_down, starmap_clust)
      tasic_purity = calcPurity(liger_down, tasic_clust)
      starmap_ari = calcARI(liger_down, starmap_clust)
      tasic_ari = calcARI(liger_down, tasic_clust)
      alignment = calcAlignment(liger_down)
      stats_starmap = rbind(stats_starmap, c(i, starmap_purity, tasic_purity, starmap_ari,  tasic_ari,alignment))
  }
  liger_seqfish = createLiger(list(tasic = tasic, seqfish = seqfish))
  liger_seqfish %<>% normalize() %>% selectGenes() %>% scaleNotCenter()
  liger_seqfish %<>% optimizeALS(k=20) %>% quantile_norm() %>% louvainCluster()
  liger_seqfish %<>% runTSNE()
  names(liger_seqfish@clusters)=rownames(liger_seqfish@tsne.coords)
  starmap_purity = calcPurity(liger_seqfish, seq_clust)
  tasic_purity = calcPurity(liger_seqfish, tasic_clust)
  starmap_ari = calcARI(liger_seqfish, seq_clust)
  tasic_ari = calcARI(liger_seqfish, tasic_clust)
  alignment = calcAlignment(liger_seqfish)
  stats_seqfish = rbind(stats, c(120, starmap_purity, tasic_purity, starmap_ari,  tasic_ari,alignment))
}
stats_starmap = stats_starmap[2:nrow(stats_starmap),]
stats_seqfish = stats_seqfish[2:nrow(stats_seqfish),]
stats_starmap = stats[order(stats_starmap[,1]),]
```

![Downsampling summary](updated_summary.PNG)
